training/heterogeneous-clusters/pt.grpc.sagemaker/hetero-pytorch-mnist.ipynb

```diff
@@ -0,0 +1,509 @@
+    "    \"iterations\": 100,  # No. of iterations in an epoch (must be multiple of 10).\n",
+    "}\n",
+    "\n",
+    "if IS_HETERO:\n",
+    "    kwargs[\"instance_groups\"] = [data_group, dnn_group]\n",
+    "    entry_point = \"launcher.py\"\n",
+    "else:\n",
+    "    kwargs[\"instance_type\"] = (\n",
+    "        \"ml.p3.2xlarge\" if IS_CLOUD_JOB else \"local\"\n",
+    "    )  # change the instance type if IS_HETERO=False\n",
+    "    kwargs[\"instance_count\"] = 1\n",
+    "    entry_point = \"train.py\"\n",
+    "\n",
+    "if IS_DNN_DISTRIBUTION:\n",
+    "    processes_per_host_dict = {\n",
+    "        \"ml.g5.xlarge\": 1,\n",
+    "        \"ml.g5.12xlarge\": 4,\n",
+    "        \"ml.p3.8xlarge\": 4,\n",
+    "        \"ml.p4d.24xlarge\": 8,\n",
+    "    }\n",
+    "    kwargs[\"distribution\"] = {\n",
+    "        \"mpi\": {\n",
+    "            \"enabled\": True,\n",
+    "            \"processes_per_host\": processes_per_host_dict[dnn_instance_type],\n",
+    "            \"custom_mpi_options\": \"--NCCL_DEBUG INFO\",\n",
+    "        },\n",
+    "    }\n",
+    "    if IS_HETERO:\n",
+    "        kwargs[\"distribution\"][\"instance_groups\"] = [dnn_group]\n",
+    "\n",
+    "    print(f\"distribution={kwargs['distribution']}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4ff19e24",
+   "metadata": {},
+   "source": [
+    "#### Step 3: Set up the Estimator\n",
+    "In order to use SageMaker to fit our algorithm, we'll create `Estimator` that defines how to use the container to train. This includes the configuration we need to invoke SageMaker training."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 55,
+   "id": "94f4c8ce",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "estimator = PyTorch(\n",
+    "    framework_version=\"1.11.0\",  # 1.10.0 or later\n",
+    "    py_version=\"py38\",  # Python v3.8\n",
+    "    role=role,\n",
+    "    entry_point=entry_point,\n",
+    "    source_dir=\"code\",\n",
+    "    volume_size=10,\n",
+    "    max_run=4800,\n",
+    "    disable_profiler=True,\n",
+    "    debugger_hook_config=False,\n",
+    "    **kwargs,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a81dcab6",
+   "metadata": {},
+   "source": [
+    "#### Step 4: Download the MNIST Data and Upload it to S3 bucket\n",
+    "\n",
+    "This is an optional step for now. The training job downloads the data on its run directly from MNIST website to the data_group node (grpc server). "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 56,
+   "id": "d0534973",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import logging\n",
+    "import boto3\n",
+    "from botocore.exceptions import ClientError\n",
+    "\n",
+    "# Download training and testing data from a public S3 bucket\n",
+    "\n",
+    "\n",
+    "def download_from_s3(data_dir=\"./data\", train=True):\n",
```

Comment:
These helper functions exist in the [Python SDK](https://sagemaker.readthedocs.io/en/stable/api/utility/s3.html).